### ISSUE TITLE
Paper - Correcting auto-differentiation in neural-ODE training

### ISSUE THREAD

**Issue by adam-hartshorne** [patrick-kidger/diffrax#265]
This may be of interest.

> Does the use of auto-differentiation yield reasonable updates to deep neural networks that represent neural ODEs? Through mathematical analysis and numerical evidence, we find that when the neural network employs high-order forms to approximate the underlying ODE flows (such as the Linear Multistep Method (LMM)), brute-force computation using auto-differentiation often produces non-converging artificial oscillations. In the case of Leapfrog, we propose a straightforward post-processing technique that effectively eliminates these oscillations, rectifies the gradient computation and thus respects the updates of the underlying flow.

https://arxiv.org/abs/2306.02192

**Comment by patrick-kidger** [comment]
This is a really nice paper -- thank you for sharing it!

At least in the weight-tied regime (=common for "true" neural ODEs, uncommon for ResNets) this seems to suggest that we can halve the cost of backpropagation when using Leapfrog -- we only really need to track the branch with non-negligible gradients.

I wonder how this problem would generalise to other linear multistep methods -- most practical usage of autodiff+diffeqsolves has involved Runge--Kutta methods.

**Comment by adam-hartshorne** [comment]
They state p3,

> We provide calculations only for Leapfrog, but the same phenomenon is expected to hold for general LMM type neural ODEs.
